Correct FFmpeg API usage

alex65536 opened the pull request:
FFmpeg docs say that you must return `AVERROR_EOF` from the read callback, not zero. Still, Telegram just propagates the return value from `IODevice::read()` call, which returns zero in case of EOF.

I don't know whether this commit has any effect on the upstream build, but it fixes a bug in Debian build of Telegram, which is using FFmpeg 5.1 instead of 4.4. Still, it's also useful in the upstream, as it makes work with FFmpeg more correct.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/telegramdesktop/tdesktop?pullRequest=25012) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/telegramdesktop/tdesktop?pullRequest=25012) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/telegramdesktop/tdesktop?pullRequest=25012) it.</sub>

darkkeks commented on Telegram/SourceFiles/media/clip/media_clip_ffmpeg.cpp:
```diff
@@ -482,7 +482,13 @@ FFMpegReaderImplementation::PacketResult FFMpegReaderImplementation::readAndProc
 int FFMpegReaderImplementation::_read(void *opaque, uint8_t *buf, int buf_size) {
 	FFMpegReaderImplementation *l = reinterpret_cast<FFMpegReaderImplementation*>(opaque);
-	return int(l->_device->read((char*)(buf), buf_size));
+	FFMpegReaderImplementation *l = reinterpret_cast<FFMpegReaderImplementation*>(opaque);
```
Seems like the line is mistakenly duplicated now
```suggestion
```

ilya-fedin commented:
> but it fixes a bug in Debian build of Telegram

Can you provide any details on the bug?

ilya-fedin commented:
> FFmpeg docs say

Can you point on that place in the docs? I don't see this documented at all.
![изображение](https://user-images.githubusercontent.com/17829319/187831944-157a6e20-7a57-4cae-9c19-70885c577670.png)

john-preston commented:
@alex65536 Hi. Thanks for your PR!

I'd like to request a couple of things. First was reviewed already, it looks like a line is duplicated now.

Second - all other uses of avio_alloc_context should be fixed as well. Those include:
- Media::AbstractFFMpegLoader::_read_data (media/audio/media_audio_ffmpeg_loader.cpp)
- Media::AbstractFFMpegLoader::_read_bytes
- Media::AbstractFFMpegLoader::_read_file
- Media::Capture::Instance::Inner::Private::_read_data (media/audio/media_audio_capture.cpp)
- Media::Streaming::File::Context::Read (media/streaming/media_streaming_file.cpp)
- FFmpeg::FrameGenerator::Impl::Read (ffmpeg/ffmpeg_frame_generator.cpp)

ilya-fedin commented:
There's no need to split such simple changes to separate commits

alex65536 commented:
> Can you provide any details on the bug?

The details are as follows. Each time when I view a specific animated sticker, the animation hangs at the very end, and the app starts consuming one CPU core until restart. If I view the same sticker three times, the app will consume three CPU cores :) That's because ffmpeg receives `0` instead of `AVERROR_EOF`, thus trying to read zero-sized buffer infinitely many times.

AFAIK the change involving `AVERROR_EOF` instead of `0` was added in FFmpeg 4.0, and the official builds seem to use 4.4. So, I don't really know why this is not reproducible on the official builds.

See also: https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=1018065. This report is not of very high quality and describes three regressions at once, and this PR definitely fixes one of them.

alex65536 commented:
> There's no need to split such simple changes to separate commits

I just pushed "Commit suggestion" in the UI :) If you prefer to keep the commits from PRs and not to squash them during merge, I may merge these two commits as one manually.

ilya-fedin commented:
> Each time when I view a specific animated sticker, the animation hangs at the very end, and the app starts consuming one CPU core until restart. If I view the same sticker three times, the app will consume three CPU cores

This was reported by other users of third-party packages, but definitely is not reproducible with any official binaries, yeah... Thanks for the fix.

But as @john-preston pointed your fix is incomplete, unfortunately.

alex65536 commented:
> But as @john-preston pointed your fix is incomplete, unfortunately.

I am going to fix the places he mentioned above soon, and then I hope this PR can be merged.

alex65536 commented:
> FFmpeg::FrameGenerator::Impl::Read (ffmpeg/ffmpeg_frame_generator.cpp)

I could not find such file, but there is `FFmpeg::EmojiGenerator::Impl::_read` instead

alex65536 commented:
Now all other uses are corrected.

john-preston commented:
@alex65536 Yeah, that's right. Thanks!

alex65536 commented:
Is anything else required for the PR to be merged?

john-preston commented:
@alex65536 Please don't rename "Read / Seek" to "_read / _seek", the second one is legacy style that I unfortunately didn't remove from the codebase :( Either you can leave the method names as they are or you can rename the legacy style to the current one (_read_file to ReadFile, etc). After that I'll merge it.

john-preston commented:
@alex65536 Also I've pushed a lot of changes, now there are the methods I've mentioned in the first comment :( Sorry for that, can you please rebase your changes.

alex65536 commented:
> Either you can leave the method names as they are or you can rename the legacy style to the current one (_read_file to ReadFile, etc).

I rebased the changes and converted all the methods to new code style.

john-preston commented:
Thanks!
